src/test/scala/is/hail/methods/LDMatrixSuite.scala

```diff
@@ -96,4 +98,130 @@ class LDMatrixSuite extends SparkSuite {
     assert(actual.toLocalMatrix == LDMatrix.read(hc, fname).toLocalMatrix)
   }
 
+  private def readCSV(fname: String): Array[Array[Double]] =
+    hc.hadoopConf.readLines(fname) { it =>
+      it.map(_.value)
+        .map(_.split(",").map(_.toDouble))
+        .toArray[Array[Double]]
+    }
+
+  private def exportImportAssert(export: (String) => Unit, expected: Array[Double]*) {
+    val fname = tmpDir.createTempFile("test")
+    export(fname)
+    assert(readCSV(fname) === expected.toArray[Array[Double]])
+  }
+
+  private def rowArrayToIRM(a: Array[Array[Double]]): IndexedRowMatrix = {
+    val rows = a.length
+    val cols = if (rows == 0) 0 else a(0).length
+    new IndexedRowMatrix(
+      sc.parallelize(a.zipWithIndex.map { case (a, i) => new IndexedRow(i, new DenseVector(a)) }),
+      rows,
+      cols)
+  }
+
+  private def rowArrayToLDMatrix(a: Array[Array[Double]]): LDMatrix = {
+    val m = rowArrayToIRM(a)
+    LDMatrix(hc, m, (0 until m.numRows().toInt).map(_ => Variant.gen.sample()).toArray[Variant], m.numCols().toInt)
+  }
+
+  @Test
+  def exportSimple() {
+    val fullExpected = Array(
+      Array(1.0, 2.0, 3.0),
+      Array(4.0, 5.0, 6.0),
+      Array(7.0, 8.0, 9.0))
+    val ldm = rowArrayToLDMatrix(fullExpected)
+
+    exportImportAssert(ldm.export(_, ",", header=None, parallelWrite=false),
+      fullExpected:_*)
+
+    exportImportAssert(ldm.exportLowerTriangle(_, ",", header=None, parallelWrite=false),
+      Array(4.0),
+      Array(7.0, 8.0))
+
+    exportImportAssert(ldm.exportStrictLowerTriangle(_, ",", header=None, parallelWrite=false),
+      Array(1.0),
+      Array(4.0, 5.0),
+      Array(7.0, 8.0, 9.0))
+
+    exportImportAssert(ldm.exportStrictUpperTriangle(_, ",", header=None, parallelWrite=false),
+      Array(1.0, 2.0, 3.0),
+      Array(5.0, 6.0),
+      Array(9.0))
+
+    exportImportAssert(ldm.exportUpperTriangle(_, ",", header=None, parallelWrite=false),
+      Array(2.0, 3.0),
+      Array(6.0))
+  }
+
+  @Test
+  def exportAllZeros() {
```

Comment:
What is the benefit of the zeros test? It seems like your test above is sufficient.

Reply:
Heh, this is a classic instance of [white-box testing](https://en.wikipedia.org/wiki/White-box_testing). I know that the implementation of the `LDMatrix` uses `IndexedRowMatrix` which doesn't even store all-zero rows. So I wanted to make sure all-zero rows are propagated in all circumstances.